samza-core/src/test/java/org/apache/samza/diagnostics/TestDiagnosticsManager.java

```diff
@@ -0,0 +1,122 @@
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import org.apache.samza.coordinator.stream.MockCoordinatorStreamSystemFactory;
+import org.apache.samza.job.model.ContainerModel;
+import org.apache.samza.metrics.reporter.MetricsSnapshot;
+import org.apache.samza.serializers.MetricsSnapshotSerdeV2;
+import org.apache.samza.system.OutgoingMessageEnvelope;
+import org.apache.samza.system.SystemStream;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.ArgumentMatcher;
+import org.mockito.Mockito;
+
+
+public class TestDiagnosticsManager {
+  private DiagnosticsManager diagnosticsManager;
+  private MockCoordinatorStreamSystemFactory.MockSystemProducer mockSystemProducer;
+  private SystemStream diagnosticsSystemStream = new SystemStream("kafka", "test stream");
+
+  private String jobName = "Testjob";
+  private String jobId = "test job id";
+  private String executionEnvContainerId = "exec container id";
+  private String taskClassVersion = "0.0.1";
+  private String samzaVersion = "1.3.0";
+  private String hostname = "sample host name";
+  private int containerMb = 1024;
+  private int numStoresWithChangelog = 2;
+  private int containerNumCores = 2;
+  private Map<String, ContainerModel> containerModels = TestDiagnosticsStreamMessage.getSampleContainerModels();
+
+  @Before
+  public void setup() {
+
+    // Mocked system producer for publishing to diagnostics stream
+    mockSystemProducer = new MockCoordinatorStreamSystemFactory.MockSystemProducer("source");
+
+    // Mocked scheduled executor service which does a synchronous run() on scheduling
+    ScheduledExecutorService mockExecutorService = Mockito.mock(ScheduledExecutorService.class);
+    Mockito.when(mockExecutorService.scheduleWithFixedDelay(Mockito.argThat(new ArgumentMatcher<Runnable>() {
+      @Override
+      public boolean matches(Object argument) {
+        Runnable runnable = (Runnable) argument;
+        try {
+          runnable.run();
+        } catch (Exception e) {
+          e.printStackTrace();
+          Assert.fail();
+        }
+        return true;
+      }
+    }), Mockito.anyLong(), Mockito.anyLong(), Mockito.eq(TimeUnit.SECONDS)))
+        .thenReturn(Mockito.mock(ScheduledFuture.class));
+
+    this.diagnosticsManager =
+        new DiagnosticsManager(jobName, jobId, containerModels, containerMb, containerNumCores, numStoresWithChangelog,
+            "0", executionEnvContainerId, taskClassVersion, samzaVersion, hostname, diagnosticsSystemStream,
+            mockSystemProducer, Duration.ofSeconds(1), mockExecutorService);
+  }
+
+  @Test
+  public void testDiagnosticsStreamPublish() {
+    // invoking start will do a syncrhonous publish to the stream because of our mocked scheduled exec service
+    this.diagnosticsManager.start();
+    Assert.assertEquals("One message should have been published", 1, mockSystemProducer.getEnvelopes().size());
```

Comment:
Should you call it twice to make sure the container info doesn't get published for the second time?

Reply:
Yup, added additional test.